src/Nethermind/Nethermind.Core/Caching/ClockKeyCacheNonConcurrent.cs

```diff
@@ -0,0 +1,136 @@
+
+public sealed class ClockKeyCacheNonConcurrent<TKey>(int maxCapacity) : ClockCacheBase<TKey>(maxCapacity)
+    where TKey : struct, IEquatable<TKey>
+{
+    private readonly Dictionary<TKey, int> _cacheMap = new();
+
+    public bool Get(TKey key)
+    {
+        if (_cacheMap.TryGetValue(key, out int offset))
+        {
+            MarkAccessedNonConcurrent(offset);
+            return true;
+        }
+        return false;
+    }
+
+    public bool Set(TKey key)
+    {
+        if (_cacheMap.TryGetValue(key, out int offset))
+        {
+            MarkAccessedNonConcurrent(offset);
+            return false;
+        }
+
+        return SetSlow(key);
+    }
+
+    private bool SetSlow(TKey key)
+    {
+        int offset = _cacheMap.Count;
+        if (FreeOffsets.Count > 0)
+        {
+            offset = FreeOffsets.Dequeue();
+        }
+        else if (offset >= MaxCapacity)
+        {
+            offset = Replace(key);
+        }
+
+        _cacheMap[key] = offset;
+        KeyToOffset[offset] = key;
+
+        return true;
+    }
+
+    private int Replace(TKey key)
+    {
+        int position = Clock;
+        int max = _cacheMap.Count;
+        while (true)
+        {
+            if (position >= max)
+            {
+                position = 0;
+            }
+
+            bool accessed = ClearAccessedNonConcurrent(position);
+            if (!accessed)
+            {
+                if (!_cacheMap.Remove(KeyToOffset[position]))
+                {
+                    ThrowInvalidOperationException();
+                }
+                break;
+            }
+
+            position++;
+        }
+
+        Clock = position + 1;
+        return position;
+
+        [DoesNotReturn]
+        static void ThrowInvalidOperationException()
+        {
+            throw new InvalidOperationException($"{nameof(ClockKeyCacheNonConcurrent<TKey>)} removing item that doesn't exist");
+        }
+    }
+
+    public bool Delete(TKey key)
+    {
+        if (_cacheMap.Remove(key, out int offset))
+        {
+            ref var node = ref KeyToOffset[offset];
+            ClearAccessedNonConcurrent(offset);
+            FreeOffsets.Enqueue(offset);
+            return true;
+        }
+
+        return false;
+    }
+
+    public new void Clear()
+    {
+        base.Clear();
+        _cacheMap.Clear();
+    }
+
+    public bool Contains(TKey key)
+    {
+        return _cacheMap.ContainsKey(key);
+    }
+
+    public int Count => _cacheMap.Count;
+
+    private bool ClearAccessedNonConcurrent(int position)
+    {
+        uint offset = (uint)position >> BitShiftPerInt64;
+        long flags = 1L << position;
+
+        ref long accessedBitmapWord = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(HasBeenAccessedBitmap), offset);
+        bool accessed = (accessedBitmapWord & flags) != 0;
+        accessedBitmapWord &= ~flags;
+
+        return accessed;
+    }
+
+    private void MarkAccessedNonConcurrent(int position)
```

Comment:
I'd push it to the the base, leave it to the caller to use.

Reply:
Done